lxd/cluster/heartbeat.go

```diff
@@ -282,20 +283,20 @@ func (g *Gateway) heartbeat(ctx context.Context, mode heartbeatMode) {
 		return
 	}
 
+	// Address of this node.
+	localAddress, err := node.ClusterAddress(g.db)
```

Comment:
Though if I'm reading things right, we're still dependent on the cluster connection for `GetNodes()` so the heartbeat round will still fail even with this change?

Reply:
Yes indeed, that was my assumption too, although it appears to be working.

Here's an example of a heartbeat loop of a formerly non-leader node being upgraded (after the leader was upgraded) - both without this patch.

https://paste.ubuntu.com/p/FJmFXmzsZW/

And you can see its sending hearbeats to itself (there's only 3 nodes):

```
t=2021-06-02T08:39:04+0000 lvl=warn msg="Failed heartbeat" address=10.98.30.9:8443 err="Failed to send heartbeat request: Put \"https://10.98.30.9:8443/internal/database\": context canceled"
t=2021-06-02T08:39:04+0000 lvl=warn msg="Failed heartbeat" address=10.98.30.6:8443 err="Failed to send heartbeat request: Put \"https://10.98.30.6:8443/internal/database\": context canceled"
t=2021-06-02T08:39:04+0000 lvl=warn msg="Failed heartbeat" address=10.98.30.5:8443 err="Failed to send heartbeat request: Put \"https://10.98.30.5:8443/internal/database\": context canceled"
```

I'm going to dig into it and figure out whats going on.

Reply:
@stgraber figured it out. When a cluster member is waiting for the other members to be upgraded it enters a connect loop with the cluster database here https://github.com/lxc/lxd/blob/966cf0739e1015acf14f21b0b5a303f881c1e80b/lxd/daemon.go#L1016-L1045

The call to `db.OpenCluster()` will return an `ErrSomeNodesAreBehind` error *but* it also returns the active database connection (my understanding here is that the actual dqlite raft cluster is up and running (schema patches aren't run until all versions match), its just the LXD level version that we are waiting to be in sync).

This is done on purpose from what I can tell, because inside the wait loop it explicitly starts up a heartbeat process (which as we know requires cluster DB access to operate).

However the `db.OpenCluster()` function is also responsible for setting the internal nodeID var using `cluster.NodeID()`, however it doesn't do that if it returns earlier with `ErrSomeNodesAreBehind`.

So the heartbeat task is started, with a nodeID var set to 0, and the formerly used `GetLocalNodeAddress()` function doesn't return an error if there is no matching node in nodes table, it just returns an empty string.

That empty string was then used to compare to the full nodes list in order to skip sending a heartbeat to itself, which ofc it didn't match, which then caused the node to send a heartbeat to itself, triggering an immediate cancellation of the ongoing heartbeat (this is the part I added recently), and to restart the heartbeat, which immediately triggered a cancellation and restart...

The loop would have continued forever, but at some point, presumably due to lack of heartbeats completing, the raft cluster demotes it from being leader and so the heartbeats stop, and none of the members are leader.

This bug has been there for a long time I think, but it only surfaced when I added heartbeat cancellation and restart on joining node notification (which is triggered when a node connects to dqlite too).